src/allmydata/testing/web.py

```diff
@@ -0,0 +1,225 @@
+            )
+        )
+    # what we do here is to start with empty hashers for the key and
+    # ueb_hash and repeatedly feed() them a zero byte on each
+    # iteration .. so the same sequence of capabilities will always be
+    # produced. We could add a seed= argument if we wanted to produce
+    # different sequences.
+    number = 0
+    key_hasher = hashlib.new("sha256")
+    ueb_hasher = hashlib.new("sha256")
+
+    # capabilities are "prefix:<128-bits-base32>:<256-bits-base32>:N:K:size"
+    while True:
+        number += 1
+        key_hasher.update("\x00")
+        ueb_hasher.update("\x00")
+
+        key = base32.b2a(key_hasher.digest()[:16])  # key is 16 bytes
+        ueb_hash = base32.b2a(ueb_hasher.digest())  # ueb hash is 32 bytes
+
+        cap = u"{kind}:{key}:{ueb_hash}:{n}:{k}:{size}".format(
+            kind=kind,
+            key=key,
+            ueb_hash=ueb_hash,
+            n=1,
+            k=1,
+            size=number * 1000,
+        )
+        yield cap.encode("ascii")
+
+
+class _FakeTahoeUriHandler(Resource, object):
+    """
+    """
+
+    isLeaf = True
+    _data = None
+    _capability_generators = None
+
+    def _generate_capability(self, kind):
+        """
+        :param str kind: any valid capability-string type
+
+        :returns: the next capability-string for the given kind
+        """
+        if self._capability_generators is None:
+            self._capability_generators = dict()
+
+        if kind not in self._capability_generators:
+            self._capability_generators[kind] = capability_generator(kind)
+        capability = next(self._capability_generators[kind])
+        return capability
+
+    def add_data(self, kind, data):
+        """
+        adds some data to our grid
+
+        :returns: a capability-string
+        """
+        assert isinstance(data, bytes)
+
+        cap = self._generate_capability(kind)
+        if self._data is None:
+            self._data = dict()
+        assert cap not in self._data, "already have '{}'".format(cap)
+        self._data[cap] = data
+        return cap
+
+    def render_PUT(self, request):
+        data = request.content.read()
+        return self.add_data("URI:CHK:", data)
+
+    def render_POST(self, request):
+        t = request.args[u"t"][0]
+        data = request.content.read()
+
+        type_to_kind = {
+            "mkdir-immutable": "URI:DIR2-CHK:"
+        }
+        kind = type_to_kind[t]
+        return self.add_data(kind, data)
+
+    def render_GET(self, request):
+        uri = DecodedURL.from_text(request.uri.decode('utf8'))
+        # XXX FIXME
+        capability = uri.query[0][1]
+
+        if self._data is None or capability not in self._data:
+            return u"No data for '{}'".format(capability).decode("ascii")
+
+        return self._data[capability]
+
+
+def create_fake_tahoe_root():
+    """
+    :returns: an IResource instance that will handle certain Tahoe URI
```

Comment:
I think this is one of the two primary (only?) public interfaces to this module so a good docstring would be nice.

I am writing a client for the webapi.  I want to test it.  How does this thing help?  What do I do with it?  What are the limitations?